cloud/ec2_util.go

```diff
@@ -804,12 +803,14 @@ func allocateIPAddressForHost(ctx context.Context, settings *evergreen.Settings,
 		return nil
 	}
 
-	allocationID, err := allocateIPAddress(ctx, c, settings.Providers.AWS.IPAMPoolID)
+	// This intentionally uses the host tag to identify the host instead of the
+	// host ID because the host ID changes after a host is created.
+	ipAddr, err := host.AssignUnusedIPAddress(ctx, h.Tag)
 	if err != nil {
 		return errors.Wrap(err, "allocating IP address")
 	}
-	if err := h.SetIPAllocationID(ctx, allocationID); err != nil {
-		return errors.Wrapf(err, "setting IP allocation ID '%s' for host", allocationID)
+	if err := h.SetIPAllocationID(ctx, ipAddr.AllocationID); err != nil {
```

Comment:
Should this be atomic with `AssignUnusedIPAddress`?

Reply:
It could be atomic but we would need a transaction because they're modifying two different collections. 

In the edge case where L808 succeeds but L812 fails, I have a follow-up PR to clean up after hosts that have mismatched state from the IP address doc (that cleanup will also handles other resource leak situations so it's not just for this one case). I imagine it would only happen here if the host got _really_ unlucky on a deploy since the two are pretty fast updates, so I'm not sure it's worth adding a transaction for such a small case.